### PR TITLE
CD-112773 Allow resource and context to be set on a pre-request basis

### PR DESCRIPTION
**Pull request by johnny-lai**
- [x] @johnwu96822

### CODE REVIEW

**Comment by johnny-lai** [comment]
Removing "WIP" because since CoupaPay doesn't actually need these changes, then I don't think we need to consider adjusting the "context" for now

File: lib/sand/service.rb [hunk 1]
```diff
@@ -103,13 +109,16 @@ def verify_token(token, options = {})
       token = token.to_s
       return {'allowed' => false} if token.empty?
 
+      resource = options.fetch(:resource, @resource)
+      raise ArgumentError.new("resource is required") if resource.nil?
```
Comment (johnwu96822) [inline lib/sand/service.rb]: Don't accept empty string as resource: `if resource.nil? || resource.empty?`

File: lib/sand/base.rb [hunk 2]
```diff
@@ -34,9 +34,13 @@ def self.cache_type
       raise NotImplementedError
     end
 
-    def cache_key(key, scopes)
-      scopes = Array(scopes).join('_')
-      ([@cache_root, self.class.cache_type, key] + (scopes.empty? ? [] : [scopes])).join('/')
+    def cache_key(key, scopes, resource)
```
Comment (johnwu96822) [inline lib/sand/base.rb]: If we default `resource = nil`, then callers who don't want to pass in resource don't need to pass in `nil` as parameter

**Comment by johnwu96822** [comment]
Minor comments

**Comment by johnny-lai** [comment]
[lockit v4.3.0] Merged via Lockit CLI on 2018-05-17 11:29:04.646616 -0700 PDT m=+4.534335788

**Pull request merged**